iothub/device/src/RetryPolicies/RetryPolicyBase.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    /// <summary>
+    /// A base retry policy.
+    /// </summary>
+    public abstract class RetryPolicyBase : IRetryPolicy
+    {
+        private readonly Random _rng = new();
+        private readonly object _rngLock = new();
+
+        /// <summary>
+        /// Creates an instance of this class.
+        /// </summary>
+        /// <param name="maxRetries">The maximum number of retries; use a negative value for infinite retries.</param>
+        protected RetryPolicyBase(uint maxRetries)
+        {
+            Argument.AssertNotNegativeValue(maxRetries, nameof(maxRetries));
+            MaxRetries = maxRetries;
+        }
+
+        /// <summary>
+        /// The maximum number of retries
+        /// </summary>
+        protected uint MaxRetries { get; private set; }
+
+        /// <inheritdoc/>
+        public virtual bool ShouldRetry(uint currentRetryCount, Exception lastException, out TimeSpan retryDelay)
+        {
+            retryDelay = TimeSpan.Zero;
+
+            return lastException is IotHubClientException hubEx
+                    && hubEx.IsTransient
+                && MaxRetries == 0
+                || currentRetryCount < MaxRetries;
+        }
+
+        /// <summary>
+        /// Gets jitter between 95% and 105% of the base time.
+        /// </summary>
+        protected TimeSpan UpdateWithJitter(double baseTimeMs)
+        {
+            // Don't calculate jitter if the value is very small
+            if (baseTimeMs < 10)
+            {
+                return TimeSpan.FromMilliseconds(baseTimeMs);
+            }
+
+            double jitterMs;
+
+            // Because Random is not threadsafe
+            lock (_rngLock)
```

Comment:
> Because Random is not threadsafe

In the sense that two calls to "rng.Next" might return the same value? If that's the only reason, I don't see the need for a lock here. I understand we'd want multiple devices to not retry with the same exact jitter, but within the scope of a single device, there is no reason I'm aware of that we need to avoid same rng values here

Reply:
If you had a gateway scenario and the same retry policy was being used with several devices, you'd risk them retrying with similar jitter values, negating the value of jitter.

I did a quick test to see how often I'd get dupes (0-99) running in a tight forloop with Task.Run:

> 33
45
6
20
86
73
16
5
83
1
3
5
33
33
38
98
93
61
64
13
97
57
57
18
67
26
57
59
5
27
74
15
62
96
56
46
70
76
95
82
75
82
87
65
83
35
40
1
82
25
82
86
72
64
39
38
92
66
49
26
44
44
98
49
18
51
83
13
32
28
17
55
88
38
97
56
98
34
19
66
37
57
52
96
88
12
36
40
46
46
8
34
70
28
63
83
96
11
35
35

I spot 4 dupes.

Reply:
So, probably not a huge deal, but this feels like something one should be more cautious about, given .NET is used for gateway.

Reply:
Adjacent dupes, I should have said.

I tried it with a lock, ran several times, only saw 1 pair of dupes in those runs. Could be statistical.

Reply:
Improved the test. Used task completion source to synchronize 1000 tries. Stored results (0 to 100k random value) in a concurrent dictionary.

With lock, had 6 dupes, all in pairs.

Without lock we get 51, the max of which is 4 dupes.

I'd say that is a significant improvement, but one could certainly argue that even for gateway, having handfuls of devices retry with the same jitter the first time around is not impactful.

I could be convinced either way. Abhipsa says it isn't complicated, so leave it in. I'm torn, but I'll leave it in for now. It won't impact single device apps in a measurable way.